### ISSUE TITLE
Attached co-services with non-unit GET params in Eliom 6

### ISSUE THREAD

**Issue by darioteixeira** [ocsigen/eliom#322]
Perhaps I'm missing something, but the definition of the type `Eliom_service.id` seems to preclude fallback services with non-unit GET params.  Is this intentional?  If so, isn't the redeclaration of GET parameters for the co-service redundant, since they have to be `Eliom_parameter.unit`?

In any case, fallbacks with non-unit GET params would be useful.  Here's one minimal example:

````ocaml
open Eliom_content.Html.F

let coucou_service = Eliom_service.create
    ~meth:(Get Eliom_parameter.(int "x"))
    ~id:(Path [])
    ()

let rec step1_handler x () =
    let step2_service =
        Eliom_registration.Html.create
            ~scope:Eliom_common.default_session_scope
            ~meth:(Post (Eliom_parameter.int "x", Eliom_parameter.string "s"))
            ~id:(Fallback coucou_service)
            step2_handler in
    let step2_form s =
        [
        Form.input ~input_type:`Text ~name:s ~value:"hello" Form.string;
        Form.button_no_value ~button_type:`Submit [pcdata "Submit"];
        ] in
    Lwt.return
        (html
        (head (title (pcdata "Main")) [])
        (body [Form.post_form step2_service step2_form x]))

and step2_handler x s =
    Lwt.return
        (html
        (head (title (pcdata "Step 2/2")) [])
        (body
            [
            p [pcdata ("X=" ^ string_of_int x)];
            p [pcdata ("S=" ^ s)];
            ]))

let () =
    Eliom_registration.Html.register coucou_service step1_handler
````

**Comment by vasilisp** [comment]
It is intentional, permitting a uniform GET = unit, POST = unit type for services.

We need to force GET = unit for all cases but `~meth:Post (_, _)`, and if I remember correctly this is hard to achieve.

**Comment by darioteixeira** [comment]
Thanks for the reply, @vasilisp. But how would you implement the example above?  Note that the GET parameter of the fallback is of type `int`.

**Comment by vasilisp** [comment]
I don't have a good answer, sorry. With the current version you would need to break the fallback relationship between `coucou_service` and `step2_service`. I can only promise to take another look at the type constraints and see if we can permit your use case.

**Comment by darioteixeira** [comment]
> I don't have a good answer, sorry. With the current version you would need to break the fallback
> relationship between coucou_service and step2_service.

But wouldn't this mean that we could only attach POST co-services to services with unit GET parameters?  There's plenty of valid use cases which don't fit this mold (I use quite a few, myself). Perhaps pre-applied services could be used to get around this limitation, but that would be an ugly hack.

> I can only promise to take another look at the type constraints and see if we can permit your use case.

Please do, because IMHO this limitation is a serious regression over the current API.

**Comment by vasilisp** [comment]
I am looking into this.

I have a hackish patch (659882b4dc3028af23b37a5c37bfda24625cf34c) that allows you to fall back to services with GET parameters. You need to use the separate `Fallback'` constructor. The attached service would need to use `Eliom_parameter.any`.

```ocaml
~meth:(Post (Eliom_parameter.any, Eliom_parameter.string "s"))
```

I am trying to merge `Fallback` and `Fallback'`, but it is tricky.

**Comment by darioteixeira** [comment]
> `~meth:(Post (Eliom_parameter.any, Eliom_parameter.string "s"))`

But `Eliom_parameter.any` defeats the purpose of using Eliom, doesn't it?

It may be better to take a step back and think about the issues introduced with the new Eliom6 API.  I still haven't dug deeply into Eliom6, but I've already encountered two serious API regressions (this one and #311).  And they both smell of something wrong with the design itself: #311 forces users to repeat themselves, and this one implies that a common use case is no longer expressable.

I reckon that there's something inconsistent with Eliom6's data model.  Now, I could lend you a hand in fixing it, but I would need to understand the rationale for such a radical change from Eliom5 to Eliom6.  Is there new functionality that wasn't expressable with Eliom5?  Or is the change motivated only by the need to simplify things internally?  While the latter is of course a worthy goal, it should not force a serious disruption of user apps and/or uglifying the API.

**Comment by vasilisp** [comment]
The old API was highly non-uniform, with a disquieting amount of boilerplate to support it. We went from 48 `*service*` functions (`{service, coservice, coservice'} * {GET, POST, PUT, DELETE} * {Http, OCaml, App, Unsafe}`) to just one, `Eliom_service.create`, and with hard-to-understand concepts like `coservice'` having mostly disappeared.

To get there, we did have to give up on some previously-expressible corner cases, and that's an acceptable trade-off. I don't think that we can substantially improve Eliom without breaking compatibility. If anything, I feel that even the new service API is too impaired by the old Eliom way of doing things. `Eliom_{service,registration,parameter}` tries to provide every possible bit of functionality, rendering nice interfaces impossible. We could instead focus on the most common 90% of the use cases, provide nice APIs for these ("less is more"), and leave the rest to low-level functions.

**Comment by darioteixeira** [comment]
> To get there, we did have to give up on some previously-expressible corner cases, and that's an acceptable trade-off. I don't think that we can substantially improve Eliom without breaking compatibility. If anything, I feel that even the new service API is too impaired by the old Eliom way of doing things. `Eliom_{service,registration,parameter}` tries to provide every possible bit of functionality, rendering nice interfaces impossible. We could instead focus on the most common 90% of the use cases, provide nice APIs for these ("less is more"), and leave the rest to low-level functions.

Thanks for clarifying the rationale for the change, @vasilisp.  I generally agree with the sentiment, and as an old (pre 1.0) Eliom user, I've been through a fair amount of API breakage.  However, it has always been possible to map old Eliom into new Eliom in a fairly straightforward manner, which is not the case with this particular issue.  And though I realise that one person's bread-and-butter is another person's corner case, I think that forcing POST co-services to have unit GET parameters is simply too constraining for real-world applications.  Case in point: suppose you have a service `edit_story` that takes the story ID as a GET parameter, and implements a wizard-like interaction for taking the user through the various steps of editing a story.  How would you even implement this cleanly in Eliom without POST co-services on a non-unit service?

Anyway, let's take a look at the fragment from the code above where the co-service is created:

```ocaml
Eliom_registration.Html.create
    ~scope:Eliom_common.default_session_scope
    ~meth:(Post (Eliom_parameter.int "x", Eliom_parameter.string "s"))
    ~id:(Fallback coucou_service)
    step2_handler
```

When the `id` is `Fallback`, it makes no sense to provide GET parameters to `meth`, because those were specified already by the fallback.  The root of the problem is that `meth` and `id` are being treated as orthogonal, but in fact they aren't.  Do you agree with this assessment?  

(Also, might be worth paging @balat and @Drup, who are deeply familiar with Eliom.)

**Comment by vasilisp** [comment]
>  How would you even implement this cleanly in Eliom without POST co-services on a non-unit service?

Technically, I guess you could use POST (non-co) services, or better use client code (with RPCs if needed) for the wizard UI steps. But I understand that this may not fit your definition of "cleanly", and that we may still want to support the old uses of coservices, and that's why I still insist on the current issue.

> The root of the problem is that `meth` and `id` are being treated as orthogonal, but in fact they aren't. Do you agree with this assessment?

You are right that they are not orthogonal, but I see this as a preexisting problem with the service API.
We have always pretended that these two dimensions give rise to a taxonomy of services; without such a mental model, Eliom services are incomprehensible. But in fact many points in the taxonomy had their own constraints derived from a weird interplay of the two dimensions, so the taxonomy collapses and the API is very confusing.

Let me try to write down the current problem a bit more formally. We want to define services based on our two dimensions (captured by the `id` and `meth` type constructors):

```ocaml
val create :
  ...                   ->
  ('g, 'p, ...) id      ->
  ('g, 'p, ...) meth    ->
  ('g, 'p, ...) service
```

`'g` and `'p` stand for the GET and POST param types; this is the bare minimum of type variables involved.

To allow for @darioteixeira uses for coservices, we could have the following definition of `id`:

```ocaml

type ( _ , _ , ... ) id =
  | Attached_get  : ( unit , unit , _ ) service -> ( _  , unit , ... ) service
  | Attached_post : ( 'g   , unit , _ ) service -> ( 'g , _    , ... ) service
  | Global        :                                ( _  , _    , ... ) service
  | Path          : ...                         -> ( _  , _    , ... ) service
```

Note the "conditional unifications" in `Attached_{get,post}`: either we leave the GET params free for the resulting service (with unit for those of the fallback), or we unify the GET params with those of the fallback. To do this "conditional" unification we need separate constructors `Attached_{get,post}`; I don't know of any other way to express a type-level "`if`". Note that we can't use the other dimension (`meth`) to express the conditional, because the type variables would "leak" into the `Global` and `Path` cases, which need to remain free.

(The `Attached_get` case could handle POST and DELETE services as well, I am leaving these out for conciseness.)

Based on the "conditional unification" trouble, I see four possibilities:

1. Implement he above scheme. It is ugly because of the `Attached_{get,post}` separation. It also suffers from a possible mismatch between the GET parameter names provided via `meth` and those of the fallback. (#311 is just about the redundancy if we only allow unit, but the problem is worse in the non-unit case where we need parameter names). We can handle the mismatch with a runtime check.

2. Prohibit GET (and POST, DELETE) attached coservices. We will have one `Attached` constructor (that has the same problem as `1.` with respect to parameter names).

3. Do what the `master` branch does: Impose unit GET parameters for the fallback. This restores symmetry and thus we can merge the `Attached_{get,post}` cases.

4. Give up on `id`, and sacrifice the unified `create`. We can have functions `create : ... -> path -> meth -> ...` (with an unaltered `meth`) and `create_global : ... -> meth -> ...`. For POST coservices, we will need a `create_attached : ... -> ('g, unit, ...) service -> ... -> 'p params -> ('g, 'p, ...) service`. We can provide an additional function for GET coservices, or prohibit these.

I am willing to implement any of the four schemes if there is consensus. But of course I may be missing something. If anybody thinks that they can cover all cases with a reasonably-clean API, I will be glad to review the PR :).

**Comment by darioteixeira** [comment]
Thanks for the write-up, @vasilisp.  There's a fair amount to digest here, and I'll come back to you later with a more detailed response.  Nevertheless, upon first inspection my preference is for solution 4:

  * It should solve both this issue and #311 in one fell swoop.
  * It allows for maximum flexibility on the user part.
  * It provides a straightforward migration path from Eliom5.

Now, I understand that after all the effort to unify `create`, this may seem like a step backwards. However, one should always keep in mind that the `create` unification is not a goal unto itself.  The goal should be to create a good, solid API.  If one comes to the conclusion that full unification is detrimental to that goal, then full unification should be ditched, regardless of how cool it is. (That old maxim "kill your darlings" always comes to mind in situations such as this...)

**Comment by vasilisp** [comment]
Option 5:

We can keep the `Path` and `Global` cases unified under the `create` function. We can rename the latter to `No_path` or something, and that gives as a single clear-enough entrypoint.

For attached co-services, we can provide a `create_attached` function as described above. Or instead we can improve what is now `attach_global_to_fallback` and simply call it `attach` (so that we always create originally-unattached coservices and attach them later).

**Comment by Drup** [comment]
Glancing at the whole thing (I haven't though about it very much), I like option 5 very much. Especially the distinction `create`/`attach`.

**Comment by vasilisp** [comment]
In fact, I think there is a very simple solution that slightly adjusts the mechanism we had in place for forcing unit GET parameters: 504005e9dc3763d7f4e16c8c4ce70d456fb0583e . Either there is something wrong with it, or I am an idiot for not seeing this sooner. That would resolve the problem, and we can do another hack for #311 (something like a `Post_star` constructor that leaves the POST parameters free).

**Comment by darioteixeira** [comment]
> In fact, I think there is a very simple solution that slightly adjusts the mechanism we had in place for forcing unit GET parameters: 504005e . Either there is something wrong with it, or I am an idiot for not seeing this sooner. That would resolve the problem, and we can do another hack for #311 (something like a Post_star constructor that leaves the GET parameters free).

Perhaps I'm missing something, but won't this force a duplicate declaration of GET parameters?

Anyway, from all the options so far, 5 is also my new favourite, as it can potentially fix this issue together with #311.  I agree that there's nothing wrong with unifying `Path` and `Global` services, since the orthogonality violation only rears its ugly head when one tries to cram `Attached` services into the unification.  That was indeed a bridge too far.  Moreover, the split between `create` and `attach` is still clean and much simpler than the proliferation of functions in Eliom < 6.

Regarding naming: `Path` and `Global` are fine descriptors.  You may consider using `Unattached` instead of `Global` for the sake of existing Eliom users, but frankly either one is fine.  I would avoid `No_path` simply on the grounds that it's generally preferable to name something by what it is instead of what it is not... ;-)

**Comment by vasilisp** [comment]
5c887d21a7e9179dd7f4006b6c998f4d51571412

**Comment by darioteixeira** [comment]
Thanks @vasilisp!

I've been testing the `service-split-att` branch, and as far as this issue and #311 are concerned, the problems are fixed.  I did notice a couple of other issues, though.  One is API related, and the other may be just an implementation bug:

  * It used to be the case that Eliom forced POST services to always have a GET fallback.  This is no longer the case.  Is it intentional?  (Good arguments can be made in favour or against this feature.)

  * POST co-services now modify the URL.  This is probably not intentional...  Here's the test sample I used:

```ocaml
open Eliom_content.Html.F

let coucou_service = Eliom_service.create
    ~meth:(Get Eliom_parameter.(int "x"))
    ~path:(Path ["coucou"])
    ()

let rec step1_handler x () =
    let step2_service =
        Eliom_registration.Html.attach_post
            ~scope:Eliom_common.default_session_scope
            ~fallback:coucou_service
            ~post_params:Eliom_parameter.(string "s")
            step2_handler in
    let step2_form s =
        [
        Form.input ~input_type:`Text ~name:s ~value:"hello" Form.string;
        Form.button_no_value ~button_type:`Submit [pcdata "Submit"];
        ] in
    Lwt.return
        (html
        (head (title (pcdata "Main")) [])
        (body [Form.post_form step2_service step2_form x]))

and step2_handler x s =
    Lwt.return
        (html
        (head (title (pcdata "Step 2/2")) [])
        (body
            [
            p [pcdata ("X=" ^ string_of_int x)];
            p [pcdata ("S=" ^ s)];
            ]))

let main_service = Eliom_service.create
    ~meth:(Get Eliom_parameter.unit)
    ~path:(Path [])
    ()

let main_handler () () =
    Lwt.return
        (html
        (head (title (pcdata "Main")) [])
        (body [p [a coucou_service [pcdata "Service 'coucou'"] 10]]))

let () =
    Eliom_registration.Html.register coucou_service step1_handler;
    Eliom_registration.Html.register main_service main_handler
```

**Comment by vasilisp** [comment]
For POST services, it is intentional (another change needed for symmetry), but to be honest it bugs me. I have a local branch with the following function:

```ocaml
(** [create_with_fallback ~fallback ~post_params ()] produces a POST
    service that shares the path and GET parameters of [fallback], and
    additionally accepts [post_params]. *)
val create_with_fallback :
  ?https:bool ->
  ?keep_nl_params:[ `All | `Persistent | `None ] ->
  ?priority:int ->
  fallback :
    ('gp, unit, get, att, non_co, non_ext, reg, 'y, 'gn, 'pn, non_ocaml)
      t ->
  post_params :
    ('pp, [ `WithoutSuffix ], 'pn) Eliom_parameter.params_type ->
  unit ->
  ('gp, 'pp, post, att, non_co, non_ext, reg, 'y, 'gn, 'pn, non_ocaml) t
```

Would you find something like this useful? I am not very happy about the proliferation of `create*` and `attach*` functions, especially if they necessitate lots of boilerplate in `Eliom_registration`. (That amounts to reverting to the old situation one step at a time, which indicates that the whole operation wasn't really worth it.)

Modifying the URL is not intentional. I will investigate. Thanks for the example!

**Comment by darioteixeira** [comment]
> For POST services, it is intentional (another change needed for symmetry), but to be honest it bugs me. I have a local branch with the following function:
> (...)
> Would you find something like this useful?

I'm divided, myself...  On the one hand, it's nice to say that Eliom ensures there's always a fallback for a POST service, but in practice there's also drawbacks to this approach (if Eliom is used to provide an API accessible via AJAX, for example, the declaration of a fallback is just a burden).

In any case, I see little benefit to `create_with_fallback`.  If the developer is prone to forget registering a fallback for every POST service, then they are also prone to forget using `create_with_fallback`.  And if the developer is careful and always makes sure there is a GET service with the same URL as a POST service, then they can use already existing Eliom infrastructure.  In either case there's nothing new that `create_with_fallback` brings to the table.  The only justification for such a function is if it were the **only** way to declare a POST service, because at least then you could claim that Eliom ensures there's **always** a fallback...

**Comment by vasilisp** [comment]
You are right. It would only offer the minor convenience of specifying the path and the GET parameters only once (for the fallback service only).

We can do type trickery to prevent the `~meth:Post (_, _) ~path:(Path _)` combination for `create` (thus making `create_with_fallback` mandatory), but things are getting messy.

**Comment by darioteixeira** [comment]
> We can do type trickery to prevent the ~meth:Post (_, _) ~path:(Path _) combination for create (thus making create_with_fallback mandatory), but things are getting messy.

Yeap, but that would also be a symptom of an orthogonality violation, which to be properly solved would require further de-unification... ;-)  Personally, I would be okay with Eliom just abandoning the enforcement of GET fallbacks for POST services.  Though it's a nice property to have in an ideal world, it does have drawbacks...

**Comment by darioteixeira** [comment]
It might be possible to have your cake and eat it too: what if Eliom checked at runtime if there were any POST services without a GET fallback and issued a warning?  I've opened ticket #357 to discuss this idea.
